parsing/intermediate/dialects/dialect1/parserImpls.ftl

```diff
@@ -4179,4 +4182,115 @@ SqlRangeNStartEnd RangeNStartEnd() :
         return new SqlRangeNStartEnd(getPos(), startLiteral, endLiteral,
             eachSizeLiteral, false, endAsterisk);
     }
-}
\ No newline at end of file
+}
+
+SqlCreateProcedure SqlCreateProcedure() :
+{
+    final Span s;
+    final SqlCreateSpecifier createSpecifier;
+    final SqlIdentifier procedureName;
+    final List<SqlCreateProcedureParameter> parameters =
+        new ArrayList<SqlCreateProcedureParameter>();
+    final CreateProcedureDataAccess access;
+    SqlLiteral numResultSets = null;
+    final CreateProcedureSecurity security;
+    final SqlNode statement;
+    SqlCreateProcedureParameter parameter;
+}
+{
+    (
+        <CREATE> { createSpecifier = SqlCreateSpecifier.CREATE; }
+    |
+        <REPLACE> { createSpecifier = SqlCreateSpecifier.REPLACE; }
+    )
+    { s = span(); }
+    <PROCEDURE>
+    procedureName = CompoundIdentifier()
+    <LPAREN>
+    [
+        parameter = SqlCreateProcedureParameter() {
+            parameters.add(parameter);
+        }
+        (
+            <COMMA>
+            parameter = SqlCreateProcedureParameter() {
+                parameters.add(parameter);
+            }
+        )*
+    ]
+    <RPAREN>
+    (
+        <CONTAINS> <SQL> { access = CreateProcedureDataAccess.CONTAINS_SQL; }
+    |
+        <MODIFIES> <SQL> <DATA> {
+            access = CreateProcedureDataAccess.MODIFIES_SQL_DATA;
+        }
+    |
+        <READS> <SQL> <DATA> {
+            access = CreateProcedureDataAccess.READS_SQL_DATA;
+        }
+    |
+        { access = CreateProcedureDataAccess.UNSPECIFIED; }
+    )
+    [
+        <DYNAMIC> <RESULT> <SETS>
+        numResultSets = UnsignedNumericLiteral() {
+            int numericNumResultSets = numResultSets.getValueAs(Integer.class);
+            if (numericNumResultSets < 0 || numericNumResultSets > 15) {
+                throw SqlUtil.newContextException(getPos(),
+                    RESOURCE.numberLiteralOutOfRange(
+                    String.valueOf(numericNumResultSets)));
+            }
+        }
+    ]
+    (
+        <SQL> <SECURITY>
+        (
+            <CREATOR> { security = CreateProcedureSecurity.CREATOR; }
+        |
+            <DEFINER> { security = CreateProcedureSecurity.DEFINER; }
+        |
+            <INVOKER> { security = CreateProcedureSecurity.INVOKER; }
+        |
+            <OWNER> { security = CreateProcedureSecurity.OWNER; }
+        )
+    |
+        { security = CreateProcedureSecurity.UNSPECIFIED; }
+    )
+    statement = CreateProcedureStmt()
+    {
+        return new SqlCreateProcedure(s.end(this), createSpecifier,
+            procedureName, parameters, access, numResultSets, security,
+            statement);
+    }
+}
+
+SqlNode CreateProcedureStmt() :
+{
+    final SqlNode e;
+}
+{
+    (
```

Comment:
nit: unnecessary parentheses at line 4273 and 4275. Or may be you are optimizing for future PRs?

Reply:
I think this is just preparing for follow-up PRs where each new control statement parsing function would be added here.

Reply:
Yeah there would be some OR statements inside these parens for the follow-up PRs

Reply:
Thanks for clarifying. That sounds good to me.